<?xml version="1.0"?>
<launch>
    <arg name="camera" default="/camera/color"/>
    <arg name="camera_depth" default="/camera/depth"/>
    <arg name="robot_height" default="1.0"/>

    <!-- https://github.com/introlab/rtabmap_ros/issues/359 -->
    <!-- SLAM Mode -> args="delete_db_on_start" and Mem/IncrementalMemory = true -->
    <node pkg="rtabmap_ros" type="rtabmap" name="rtabmap" args="--delete_db_on_start">

        <param name="Mem/IncrementalMemory" value="true"/>  <!-- SLAM Mode -->

        <!-- INPUT -->
        <param name="subscribe_depth"       value="false"/>
        <param name="subscribe_rgbd"        value="true"/>
        <param name="subscribe_rgb"         value="false"/>
        <param name="subscribe_scan_cloud"  value="true"/>
        <param name="subscribe_scan"        value="false"/>
        <param name="approx_sync"           value="true"/>

        <remap from="rgbd_image"    to="rgbd_image"/>
        <remap from="scan_cloud"    to="$(arg camera_depth)/points"/>
        <remap from="imu"           to="/imu/data"/>

        <!-- OUTPUT -->
        <remap from="grid_map" to="/map"/>                          <!-- Input for the global costmap -->
        <!-- <remap from="octomap_obstacles" to="/octomap_obstacles"/>  --> <!-- PointCloud2 output can be used in obstacle detection --> 

        <!-- FRAMES -->
        <param name="frame_id"      value="base_link"/>
        <param name="odom_frame_id" value="odom"/>
        <param name="map_frame_id"  value="map"/>
        <param name="publish_tf"    value="true"/>

        <!-- PARAMETERS -->
        <param name="queue_size"        value="10"/>
        <param name="wait_imu_to_init"  value="true"/>  <!-- add imu for orientation -->

        <!-- Rtabmap Parameters -->
        <param name="Grid/3D"                           value="true"/>  <!-- generate octomap -->
        <param name="Grid/FromDepth"                    value="false"/> <!-- occupancy grid from 3D lidar and NOT from depth image -->
        <param name="Grid/GroundIsObstacle"             value="false"/>
        <param name="Grid/RangeMax"                     value="5.0"/>   <!-- Maximum range from sensor. 0=inf. -->
        <param name="Grid/RayTracing"                   value="true"/>
        <param name="Grid/CellSize"                     value="0.05"/>  <!-- resolution of the occupancy grid -->
        <param name="Grid/MaxObstacleHeight"            value="$(arg robot_height)"/>   <!-- max obstacle height should be little higher than robot's height in order to have correct projection -->
        <param name="RGBD/NeighborLinkRefining"         value="true"/>
        <param name="RGBD/ProximityBySpace"             value="false"/>
        <param name="RGBD/AngularUpdate"                value="0.05"/>
        <param name="RGBD/LinearUpdate"                 value="0.05"/>
        <param name="RGBD/OptimizeFromGraphEnd"         value="false"/>
        <param name="RGBD/ProximityPathMaxNeighbors"    value="10"/>    <!-- https://github.com/introlab/rtabmap_ros/issues/270 -->
        <param name="Reg/Force3DoF"                     value="true"/>
        <param name="Reg/Strategy"                      value="1"/>     <!-- Strategy used for NeighborLinkRefining | 0=VO, 1=ICP, 2=VO+ICP | https://github.com/introlab/rtabmap_ros/issues/270 -->
        <param name="Rtabmap/StartNewMapOnLoopClosure"  value="true"/>  <!-- http://wiki.ros.org/rtabmap_ros/Tutorials/Advanced%20Parameter%20Tuning#Odometry_Auto-Reset -->
<?ignore
        <!-- Robust Graph Optimization : https://github.com/introlab/rtabmap/wiki/Robust-Graph-Optimization#ros https://github.com/introlab/rtabmap_ros/issues/433 -->
        <param name="Optimizer/Strategy"    value="2"/> <!-- 0=TORO, 1=g2o, 2=GTSAM and 3=Ceres -->
        <param name="Optimizer/Robust"      value="true"/>
        <param name="RGBD/OptimizeMaxError" value="0"/> <!-- should be 0 if Optimizer/Robust is true -->
?>
        <!-- GPS integration : https://answers.ros.org/question/329165/using-rtabmap_ros-with-external-gps/ -->
        <param name="Rtabmap/LoopGPS"   value="true"/>
        <param name="RGBD/LocalRadius"  value="20"/>
        <remap from="gps/fix"  to="/navsat/fix"/>

        <!-- ICP parameters -->
        <param name="Icp/PointToPlane"              value="false"/>
        <param name="Icp/VoxelSize"                 value="0.05"/>
        <param name="Icp/MaxCorrespondenceDistance" value="0.1"/>
    </node>

<?ignore
    <node pkg="rtabmap_ros" type="rtabmapviz" name="rtabmapviz">
        <param name="subscribe_depth"       value="false"/>
        <param name="subscribe_rgbd"        value="true"/>
        <param name="subscribe_scan_cloud"  value="true"/>
        <param name="subscribe_scan"        value="false"/>
        <param name="subscribe_odom_info"   value="false"/>

        <remap from="rgbd_image"    to="rgbd_image"/>
        <remap from="scan_cloud"    to="$(arg camera_depth)/points"/>

        <param name="frame_id"      value="base_link"/>
        <param name="odom_frame_id" value="odom"/>
    </node>
?>

</launch>
